Autobackup enhancements + support autobackup in PS

UdjinM6 opened the pull request:
Autobackup refactoring and improvements:
- make `nWalletBackups` globally accessible;
- move autobackup code from init.cpp to walletdb.cpp, see `AutoBackupWallet` function;
- refactor autobackup code to warn user if autobackup failed instead of silently ignoring this fact;
- refactor autobackup code to be able to backup fresh new wallet right after it was created, add this functionality to init sequence;
- add new cmd-line option `-walletbackupsdir` to specify full path to directory for automatic wallet backups, see `GetBackupsDir` function.

Autobackup support in PS:
- add `nKeysLeftSinceAutoBackup` to have some idea how many keys in keypool are more or less safe, show it in advanced PS UI mode and in rpc output for `privatesend` and `getwalletinfo` commands;
- add autobackups support in PrivateSend mixing both in daemon and QT mode, warn user if number of keys left since last autobackup is very low or even stop mixing completely if it's too low.

crowning- commented:
LOVE :heart: this one https://github.com/UdjinM6/dash/blob/autobackupEnhancement/src/qt/overviewpage.cpp#L165

Will test/review later...

crowning- commented:
What's fCreateAutoBackups for?

UdjinM6 commented:
@crowning- Hint: for testing purposes you can just bump PS_KEYS_THRESHOLD_WARNING and PS_KEYS_THRESHOLD_STOP to smth like 900 and 850 respectively ;)

crowning- commented on src/wallet/walletdb.cpp:
```diff
@@ -904,6 +908,113 @@ bool BackupWallet(const CWallet& wallet, const string& strDest)
+            strWalletFile = wallet->strWalletFile;
+            fs::path backupFile = backupsDir / (strWalletFile + dateTimeStr);
+            if(!BackupWallet(*wallet, backupFile.string())) {
+                strBackupWarning = strprintf(_("Failed to create backup %s!"), backupFile.string());
```
I'd add a check if the file already exists, because if it already exists (means the wallet was re-started in < 60 secs) it would make sense to either silently overwrite the existing one or do nothing.
In either case I wouldn't create a warning popup (it's uncritical because we have a recent backup in both cases). It will just confuse users...

![backup_error](https://cloud.githubusercontent.com/assets/10080039/16037377/7aec1346-3220-11e6-8b56-f410cc278c3f.jpg)

UdjinM6 replied:
There is another use case however (which fits in frequent restart case much more then what user usually do imo) - you replaced wallet.dat with another one and you think you have an auto backup (wallet gave you no warning) but you don't. Overwriting is not an option in this case either imo. I'd rather warn user all the way then answer "why it didn't work?". Maybe we can warn in a more verbose way (giving more description) though.

crowning- commented:
It would be nice to automatically stop mixing after
![lowkey](https://cloud.githubusercontent.com/assets/10080039/16037973/3513805e-3223-11e6-8179-4d55083a06c7.jpg)

because I just had to click that popup away...about 50 times...

UdjinM6 commented:
re popup: hmm... this shouldn't be the case... I might have messed up smth, will have a look

UdjinM6 commented:
@crowning- I'm not able to reproduce this on mac and I'm not quite sure how you get 50 msg boxes at once - they are meant to be modal dialogs so UI thread should pause waiting for user input afaik... Or are these messages not actually modal dialogs on Linux? 0_o

crowning- commented:
Not at once, one after another.
As soon as I confirm one the next one pops up.

UdjinM6 commented:
Added commit for more descriptive warning in "fast restart" case. Also it's more informative and less restrictive now - autobackups are enabled in this case unlike for other file system errors on `copy`.

crowning- commented:
This (in `privateSendStatus()`) https://github.com/UdjinM6/dash/blob/autobackupEnhancement/src/qt/overviewpage.cpp#L475
fails when the remaining keys are < PS_KEYS_THRESHOLD_WARNING
and this https://github.com/UdjinM6/dash/blob/autobackupEnhancement/src/qt/overviewpage.cpp#L178
triggers it over and over again. Popups till the end of time....

When we get the PS_KEYS_THRESHOLD_WARNING it should, after the first popup, be either ignored until PS_KEYS_THRESHOLD_STOP applies, or, even better, generate automatically new keys .

crowning- commented:
Okay, I found the reason: when you START a wallet which already has < PS_KEYS_THRESHOLD_WARNING keys you run into this problem.
If your keys are decreased while the wallet is running everything works as designed :thumbsup: , so this will mostly happen to upgrading users.

Add a check for PS_KEYS_THRESHOLD_WARNING during startup of the wallet (before the timer is started), add new keys when needed and everything should be fine.

UdjinM6 commented:
I was able to reproduce "endless popup" issue only for locked wallets so I took a slightly different approach:
- refactored to address locked wallets issue;
- keypool should be replenished and autobackup should be re-initialized on wallet unlock now but only if autobackup was disabled due to keypool replenishment issue.

I also adjusted few message strings (dialog boxes should look slightly less messy) and added a comment describing `nWalletBackups` values.

crowning- commented:
FYI, my testwallet was unlocked all the time. I'll check your changes after work.

To re-produce:
- use an older wallet (unlocked, no pwd) and bring the keypool down to e.g. 950.
- build your new wallet with PS_KEYS_THRESHOLD_WARNING = 970 and start it.
- enjoy :-)

UdjinM6 commented:
@crowning- I wonder how you were able to
> bring the keypool down to e.g. 950.

it always replenishes itself if wallet is unlocked...

crowning- commented:
> I wonder how you were able to bring the keypool down to e.g. 950.

Easy: Start mixing, then Reset -> Try Mix -> Resest -> Try Mix -> ...

But with your last commit everything works as designed, when I stop the wallet with e.g. 950 keys left it ups the keys to 1000 during the next start. It did not do this with your very first version.

ACK from my side...:thumbsup:

schinzelh commented:
utACK, squashed and merged
